File: Vectors/Python/vector.py
import math

class Vector:
    def __init__(self, items):
        if len(items) != 3:
            raise ValueError("A vector must have exactly 3 components.")
        self.items = items

    def __add__(self, other):
        return Vector([a + b for a, b in zip(self.items, other.items)])

    def __sub__(self, other):
        return Vector([a - b for a, b in zip(self.items, other.items)])

    def scale(self, scalar):
        return Vector([scalar * x for x in self.items])

    def dot_product(self, other):
        return sum(a * b for a, b in zip(self.items, other.items))

    def magnitude(self):
        return math.sqrt(sum(x ** 2 for x in self.items))

    def cross_product(self, other):
        return Vector([
            self.items[1] * other.items[2] - self.items[2] * other.items[1],
            self.items[2] * other.items[0] - self.items[0] * other.items[2],
            self.items[0] * other.items[1] - self.items[1] * other.items[0]
        ])

    def angle(self, other):
        dot = self.dot_product(other)
        magnitudes = self.magnitude() * other.magnitude()
        if magnitudes == 0:
            raise ValueError("Cannot calculate the angle with a zero vector.")
        cos_theta = max(-1.0, min(1.0, dot / magnitudes))
        return math.acos(cos_theta)

    def __str__(self):
        return f"Vector({self.items})"

def read_vector():
    while True:
        try:
            items = list(map(int, input("Enter 3 integers separated by spaces: ").split()))
            if len(items) != 3:
                raise ValueError("The vector must have exactly 3 components.")
            return Vector(items)
        except ValueError as e:
            print(f"Invalid input: {e}")

File: Vectors/C++/vector.cpp
#include "vector.h"

Vector::Vector(const std::vector<double>& components) {
    if (components.size() != 3) {
        throw std::invalid_argument("A vector must have exactly 3 components.");
    }
    items = components;
}

Vector Vector::operator+(const Vector& other) const {
    return Vector({ items[0] + other.items[0], items[1] + other.items[1], items[2] + other.items[2] });
}

Vector Vector::operator-(const Vector& other) const {
    return Vector({ items[0] - other.items[0], items[1] - other.items[1], items[2] - other.items[2] });
}

Vector Vector::scale(double scalar) const {
    return Vector({ items[0] * scalar, items[1] * scalar, items[2] * scalar });
}

double Vector::dotProduct(const Vector& other) const {
    return items[0] * other.items[0] + items[1] * other.items[1] + items[2] * other.items[2];
}

double Vector::magnitude() const {
    return std::sqrt(items[0] * items[0] + items[1] * items[1] + items[2] * items[2]);
}

Vector Vector::crossProduct(const Vector& other) const {
    return Vector({
        items[1] * other.items[2] - items[2] * other.items[1],
        items[2] * other.items[0] - items[0] * other.items[2],
        items[0] * other.items[1] - items[1] * other.items[0]
        });
}

double Vector::angle(const Vector& other) const {
    double dot = dotProduct(other);
    double magnitudes = magnitude() * other.magnitude();
    if (magnitudes == 0) {
        throw std::invalid_argument("Cannot calculate the angle with a zero vector.");
    }
    return std::acos(std::max(-1.0, std::min(1.0, dot / magnitudes)));
}

bool Vector::operator==(const Vector& other) const {
    return items == other.items;
}

void Vector::print() const {
    std::cout << "Vector(" << items[0] << ", " << items[1] << ", " << items[2] << ")";
}
